docs/scope-and-objectives.md

```diff
@@ -0,0 +1,168 @@
+- __Workload cluster__: A cluster whose lifecycle is managed by the Management cluster.
+- __Operating system__: A generically understood combination of a kernel and system-level userspace interface, such as Linux or Windows, as opposed to a particular distribution.
+- __Manage a cluster__: create, scale, upgrade, destroy.
+
+## Motivation
+
+Kubernetes has a common set of APIs (see the [Kubernetes API Conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md)) to orchestrate containers regardless of deployment mechanism or cloud provider. Kubernetes also has APIs for handling some infrastructure, like load-balancers, ingress rules, or persistent volumes, but not for creating new machines.
+As a result, the deployment mechanisms that manage Kubernetes clusters each have unique APIs and implementations for how to handle lifecycle events like cluster creation or deletion, control plane upgrades, and node upgrades.
+
+### Goals
+
+- To manage the lifecycle (create, scale, upgrade, destroy) of Kubernetes-conformant clusters using a declarative API.
+- To work in different environments, both on-premises and in the cloud.
+- To define common operations, provide a default implementation, and provide the ability to swap out implementations for custom ones.
+- To reuse and integrate existing ecosystem components rather than duplicating their functionality (e.g. node-problem-detector, cluster autoscaler, SIG-Multi-cluster).
+- To provide a transition path for Kubernetes lifecycle products to adopt Cluster API incrementally. Specifically, existing cluster lifecycle management tools should be able to adopt Cluster API in a staged manner, over the course of multiple releases, or even adopting a subset of Cluster API.
+
+### Non-goals
+
+- To add these APIs to Kubernetes core (kubernetes/kubernetes).
+   -  This API should live in a namespace outside the core and follow the best practices defined by api-reviewers, but is not subject to core-api constraints.
+- To manage the lifecycle of infrastructure unrelated to the running of Kubernetes-conformant clusters.
+- To force all Kubernetes lifecycle products (kops, kubespray, GKE, AKS, EKS, IKS etc.) to support or use these APIs.
+- To manage non-Cluster API provisioned Kubernetes-conformant clusters.
+- To manage a single cluster spanning multiple infrastructure providers.
+- To configure a machine at any time other than create or upgrade.
+- To duplicate functionality that exists or is coming to other tooling, e.g., updating kubelet configuration (c.f. dynamic kubelet configuration), or updating apiserver, controller-manager, scheduler configuration (c.f. component-config effort) after the cluster is deployed.
+
+## Requirements
+
+#### Foundation
+
+- Cluster API MUST be able to be deployed and run on Kubernetes.
+
+- Cluster API MUST provide isolation between providers.
```

Comment:
Providers should not be able to access the internal state or secrets of other providers.

I don't think there should be anything to prevent code sharing, the primary concern was to avoid a 'rogue' provider from stealing secrets from other providers.

Reply:
Great, I concur.  Maybe this could be `Cluster API MUST provide runtime isolation between providers within the same Management Cluster`

Reply:
@detiber's explanation could be used as the requirement: "Cluster API MUST provide state isolation between the providers on the same management cluster".